sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,550 @@
+   */
+  kEpmpEntryConfigureBadRegion,
+
+  /**
+   * The transaction is in a bad state. This can be caused by:
+   *  * Missing call the `epmp_transaction_start`.
+   *  * An attempt to configure more entries than was initially specified.
+   *  * A prior configuration attempt encountered an error.
+   */
+  kEpmpEntryConfigureBadTransaction,
+
+  /**
+   * The requested entry is out of range or has previously been configured
+   * in this transaction. Each entry may only be configured once per
+   * transaction.
+   */
+  kEpmpEntryConfigureBadEntry,
+
+  /**
+   * Encoding the entry would interfere with a different pre-existing entry.
+   *
+   * New entries will be rejected if they:
+   *  * Modify the start or end address of an adjacent TOR entry.
+   *  * Would result in an address being used in both a NAPOT/NA4 entry and a
+   *    TOR entry.
+   */
+  kEpmpEntryConfigureConflict,
+} epmp_entry_configure_result_t;
+
+/**
+ * Disable EPMP address matching.
+ *
+ * Address matching is disabled for `entry`. The pmpaddr for `entry` is set
+ * to the value of `region.start` which must also match `region.end` (i.e.
+ * the length of the region should be 0).
+ *
+ * IMPORTANT: the entry will not be updated until `epmp_transaction_end`
+ * is called.
+ *
+ * Example:
+ *
+ *   ...
+ *   res0 = epmp_transaction_start(kEpmpEntryCount2);
+ *   res1 = epmp_entry_configure_off(kEpmpEntry0,
+ *              (epmp_region_t){0},
+ *              kEpmpPermUnlocked);
+ *   res2 = epmp_entry_configure_off(kEpmpEntry1,
+ *              (epmp_region_t){ .start = 0x10, .end = 0x10 },
+ *              kEpmpPermLockedNoAccess);
+ *   res3 = epmp_transaction_end(kEpmpEntryCount2);
+ *   ...
+ *
+ * Result:
+ *
+ *   Entry | Value of `pmpaddr` | Value of `pmpcfg` |
+ *   ======+====================+===================+
+ *       0 |   0x00 (0x00 >> 2) |         0b0000000 |
+ *       1 |   0x04 (0x10 >> 2) |         0b1000000 |
+ *
+ * @param entry Entry index to update (0 <= `entry` < `kEpmpNumRegions`)
+ * @param region Updated value for pmpaddr (see above).
+ * @param permissions Updated permissions to write to pmpcfg for `entry`.
+ * @return `epmp_entry_configure_result_t`.
+ */
+epmp_entry_configure_result_t epmp_entry_configure_off(epmp_entry_t entry,
+                                                       epmp_region_t region,
+                                                       epmp_perm_t permissions);
+
+/**
+ * Configures EPMP address matching to Top Of Range (TOR).
+ *
+ * Address matching is configured as TOR for the `entry` provided.
+ *
+ * The `region` end address will be written to the `pmpaddr` for `entry`.
+ *
+ * The `region` start address will be written to the `pmpaddr` for the
+ * entry preceding `entry` if it is disabled (i.e. set to OFF). If
+ * the preceding entry is set to TOR then the start address must match
+ * the pre-existing `pmpaddr` for that entry (or 0 if `entry` is 0).
+ * All other configurations will be rejected.
```

Comment:
> If the preceding entry is set to TOR then the start address must match the pre-existing `pmpaddr`. 

nit: I suspect this is to support TOR range extension. I recommend to add a note to explain this assumption. This will make follow up reviews easier.